src/compiler/utilities.ts

```diff
@@ -2056,19 +2056,51 @@ namespace ts {
             }
             return AssignmentDeclarationKind.ObjectDefinePropertyValue;
         }
-        if (expr.operatorToken.kind !== SyntaxKind.EqualsToken ||
-            !isPropertyAccessExpression(expr.left)) {
+        if (expr.operatorToken.kind !== SyntaxKind.EqualsToken) {
             return AssignmentDeclarationKind.None;
         }
-        const lhs = expr.left;
-        if (isEntityNameExpression(lhs.expression) && lhs.name.escapedText === "prototype" && isObjectLiteralExpression(getInitializerOfBinaryExpression(expr))) {
-                // F.prototype = { ... }
-                return AssignmentDeclarationKind.Prototype;
+        if (isPropertyAccessExpression(expr.left) || isElementAccessExpression(expr.left)) {
+            const lhs = expr.left;
+            if (isEntityNameExpression(lhs.expression) && getElementOrPropertyAccessName(lhs) === "prototype" && isObjectLiteralExpression(getInitializerOfBinaryExpression(expr))) {
+                    // F.prototype = { ... }
+                    return AssignmentDeclarationKind.Prototype;
+            }
+            return getAssignmentDeclarationPropertyAccessKind(lhs);
+        }
+        return AssignmentDeclarationKind.None;
+    }
+
+    // TODO: Signed numeric literals?
+    /* @internal */
+    export function getElementOrPropertyAccessImmediateRHS(node: PropertyAccessExpression | ElementAccessExpression): Identifier | StringLiteralLike | NumericLiteral | ElementAccessExpression | undefined {
+        if (isPropertyAccessExpression(node)) {
+            return node.name;
+        }
+        const arg = skipParentheses(node.argumentExpression);
+        if (isNumericLiteral(arg) || isStringLiteralLike(arg)) {
+            return arg;
+        }
+        return node;
+    }
+
+    /* @internal */
+    export function getElementOrPropertyAccessName(node: PropertyAccessExpression | ElementAccessExpression): string | undefined {
+        const name = getElementOrPropertyAccessImmediateRHS(node);
+        if (name) {

```

Comment:
I thought the code from here until the end of the function was already in some other utility

Reply:
Looks like not. Maybe it's stuck in the checker somewhere?

Reply:
I don't think we've had something quite like this before - it's supposed to return the name if the name is able to be interpreted as not late bound, and `undefined` otherwise.